### PR TITLE
Add ttnn.sampling op for fused non-greedy LLM sampling

### PR DESCRIPTION
**Pull request by kmabeeTT**
## Ticket

tenstorrent/tt-xla#3940

## Problem

Non-greedy (temperature > 0) LLM inference on TT device was significantly slower than greedy. The sampling path (softmax → top-k filtering → top-p filtering → multinomial sampling) ran as separate compiled ops over a pre-filtered candidate set of ~128 tokens. Each op incurs program dispatch overhead and separate kernel launches.

The `ttnn::sampling` hardware kernel fuses all of these into a single call: softmax + top-k masking + top-p masking + multinomial selection. Exposing it through the compiler stack eliminates those intermediate kernel launches.

## What's changed

Wires the existing `ttnn::sampling` hardware kernel end-to-end through the tt-mlir compiler and runtime:

**Compiler:**
- Add `ttir.sampling` op to the TTIR dialect with verifier (inputs: candidate logit values, global vocab indices, per-request k/p/temp tensors, optional seed; output: sampled global token index per request)
- Add `ttnn.sampling` op to the TTNN dialect with layout workarounds (TILE for input values, ROW_MAJOR for index/param tensors and for the output — kernel produces ROW_MAJOR, so the workaround pass now inserts an explicit `ttnn.to_layout` after the op rather than having the runtime silently convert)
- Lower `stablehlo.custom_call "tt.sampling"` → `ttir.sampling` → `ttnn.sampling` through the standard conversion pipeline; optional seed parsed from `mhlo.frontend_attributes`
- Flatbuffer serialization via `reduction.fbs`

**Runtime:**
- `runtime/lib/ttnn/operations/reduction/sampling.cpp`: dispatch `SamplingOp` to `::ttnn::sampling` with required shape/layout/dtype workarounds (2D→4D reshape for batch=32 kernel requirement, defensive ROW_MAJOR enforcement, int32→uint32 k typecast, uint32→int32 output typecast)
- Wire into `program_executor.cpp` dispatch table and `runtime.cpp`

**OpModel:**
- Full `OpModel<SamplingOp>` template specialization in `TTNNOpModel.cpp` (5-tensor `convertToTensorSpec` + `QUERY_OP_CONSTRAINTS/RUNTIME(::ttnn::sampling, ...)`)
- `TTNNOpModelInterface.cpp` wires `SamplingOp::getOpConstraints/getOpRuntime` to the `OpModel` (replacing the `MissingMetalDefinition` stubs)

**Tests:**
- `test/ttmlir/Conversion/StableHLOToTTIR/sampling_op.mlir`: custom_call lowering with and without seed
- `test/ttmlir/Dialect/TTIR/reduction/sampling_tests_positive.mlir`: TTIR op round-trip
- `test/ttmlir/Dialect/TTNN/Transforms/Workarounds/sampling_workaround.mlir`: TILE → ROW_MAJOR layout conversion for index/param tensors
- `test/unittests/OpModel/TTNN/Op/TestOpModelInterface.cpp`: `OpModelBase.SamplingOp` unit test (batch=32, candidates=128)
- `tools/builder/ttnn/ttnn_builder.py` + `test/python/golden/test_ttnn_ops.py`: `TTNNBuilder.sampling()` method and `test_sampling` parametrized over Llama-3.x and OPT-125M vocab sizes (PCC check disabled because multinomial sampling is stochastic; test verifies compile + device execute succeed end-to-end)

## Results

Drops non-greedy sampler from ~34ms to ~20ms in llama cases.

Measured on Blackhole (single device, batch=1, bfp8, opt-level-0):

| Model | Non-greedy before | Non-greedy with ttnn.sampling |
|---|---|---|
| Llama 3.1-8B | 9.2 tok/s | **10.8 tok/s (+17%)** |
| Llama 3.2-1B | 14.0 tok/s | **17.2 tok/s (+23%)** |
| OPT-125M | 33.5 tok/s | **40.1 tok/s (+20%)** |

The baseline (9.2/14.0/33.5) already includes a separate tt-xla change (tenstorrent/tt-xla#4334) that replaces full-vocab sort with chunked topk. Combined improvement vs the original sort-based baseline (4.6/5.5/12.9 tok/s) is 2-3x. Output quality verified on Llama 3.1-8B and 3.2-1B.

Further throughput improvements are in progress: pre-padding the logit batch to 32 before the topk stage removes extra padding ops at the sampling layer and is expected to add another ~17%.

## Checklist

- [x] ttnn.sampling op wired end-to-end through compiler and runtime
- [x] OpModel implemented with unit test
- [x] Builder API method + builder test added
- [x] MLIR lit tests added (conversion, TTIR dialect, workarounds)
- [x] Throughput improvement verified on 3 models

### CODE REVIEW

**Comment by kmabeeTT** [comment]
special thanks to @svuckovicTT for .claude/skills/add-op/SKILL.md being added recently - claude pretty much one-shotted this and the tt-xla integration, and it worked for me without too much trouble.

**Review by mtopalovicTT — APPROVED** [review]

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 1]
```diff
@@ -7362,6 +7362,67 @@ class StableHLOPagedUpdateCacheConversionPattern
     return success();
   }
 };
+
+class StableHLOSamplingConversionPattern
```
Comment (sdjordjevicTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:7366]: Do we have sampling lowered from frontends as a custom call or composite?

File: lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp [hunk 2]
```diff
@@ -3634,6 +3634,24 @@ PagedFillCacheOp::getOpRuntime(const std::vector<TTNNLayoutAttr> &inputs,
...
       batchIdxShape, batchIdxLayout, opConfig.outputLayout);
 }
+//===----------------------------------------------------------------------===//
+// SamplingOp - TTNN Op Model Interface
+//===----------------------------------------------------------------------===//
+
+llvm::Expected<op_model::OpConstraints>
+SamplingOp::getOpConstraints(const std::vector<TTNNLayoutAttr> &inputs,
+                             const OpConfig &opConfig) {
+  return issueErrorForGetOpConstraints(
```
Comment (sdjordjevicTT) [inline lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp:3644]: This is a single-device operation if I am following correctly. If that is true, we should implement op model.

File: test/ttmlir/Dialect/TTNN/Transforms/Workarounds/sampling_workaround.mlir [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
```
Comment (sdjordjevicTT) [inline test/ttmlir/Dialect/TTNN/Transforms/Workarounds/sampling_workaround.mlir:1]: One builder test would be great.

Comment (jameszianxuTT) [reply test/ttmlir/Dialect/TTNN/Transforms/Workarounds/sampling_workaround.mlir:1]: Yes, please. We should be ensuring op accuracy on tt-mlir side to ensure no mistakes are made in the lowering.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 1 above)
Comment (kmabeeTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:7366]: We use stablehlo.custom_call @tt.sampling — not composite. In python_package/tt_torch/custom_ops.py, the op is registered as torch.library.custom_op("tt::sampling") and XLA dispatches it via stablehlo_custom_call. The seed is passed as a frontend_attribute.

(Edit some more thinking: We could switch to composite, but that would require adding a StableHLO decomposition. Since this is a TT-specific fused kernel, custom call is correct for now.)

Comment (sdjordjevicTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:7366]: Great, thanks for clarification!

File: lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp (hunk 2 above)
Comment (kmabeeTT) [reply lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp]: Hi @sdjordjevicTT  - Taking a look... though I wondered also in interest in de-risking this, can it be a follow up task after tt-deploy maybe since it works for me in tt-inference-server without the op model? Currently uses stub like many other ops in lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp that don't have op model implemented.

**Comment by jgrimTT** [comment]
Just checking before I review, was the decision to leave the ttir.sampling op out of the builder ecosystem intentional?

**Comment by kmabeeTT** [comment]
> Just checking before I review, was the decision to leave the ttir.sampling op out of the builder ecosystem intentional?

thx for looking - maybe not intentional @jgrimTT - though a builder test was flagged earlier as missing by Stefan, and I just pushed changes to address all prior feedback including builder test (https://github.com/tenstorrent/tt-mlir/pull/8043/commits/85c269d685e83dc32b6645e3616ed03b9fc23fdc) is that what you were thinking of?

```
pytest test/python/golden/test_ttnn_ops.py -k sampling -v |& tee sampling_builder_tests.log
================================================================================================================================================================================ test session starts =================================================================================================================================================================================
platform linux -- Python 3.12.3, pytest-9.0.2, pluggy-1.6.0 -- /opt/ttmlir-toolchain/venv/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.12.3', 'Platform': 'Linux-6.8.0-57-generic-x86_64-with-glibc2.39', 'Packages': {'pytest': '9.0.2', 'pluggy': '1.6.0'}, 'Plugins': {'forked': '1.6.0', 'json-report': '1.5.0', 'metadata': '3.1.1'}}
rootdir: /localdev/kmabee/tt-mlir/test/python/golden
configfile: pytest.ini
plugins: forked-1.6.0, json-report-1.5.0, metadata-3.1.1
collecting ... collected 1488 items / 1486 deselected / 2 selected

test/python/golden/test_ttnn_ops.py::test_sampling[opt] PASSED                                                                                                                                                                                                                                                                                                                 [ 50%]
test/python/golden/test_ttnn_ops.py::test_sampling[llama] PASSED                                                                                                                                                                                                                                                                                                               [100%]
Closing device for end of session
```

**Comment by kmabeeTT** [comment]
## Update: OpModel, builder test, and output layout workaround

  Added in response to review feedback — ready for another pass:

  - **OpModel:** full `OpModel<SamplingOp>` implementation with 5-tensor `convertToTensorSpec` + `QUERY_OP_CONSTRAINTS(::ttnn::sampling, ...)`; `OpModelInterface` now wires `getOpConstraints`/`getOpRuntime` (replacing the `MissingMetalDefinition` stubs). Unit test `OpModelBase.SamplingOp` added in `TestOpModelInterface.cpp` (batch=32, candidates=128).
  - **Builder test:** `TTNNBuilder.sampling()` method + `test_sampling` (parametrized over Llama-3.x and OPT-125M vocab sizes). PCC check disabled since multinomial sampling is stochastic; the test verifies compile + device execute end-to-end.
  - **Output layout workaround:** declared `ttnn.sampling` output as `ROW_MAJOR` in the workaround factory. The kernel returns `ROW_MAJOR`, so the pass now inserts an explicit `ttnn.to_layout(tile)` after the op rather than having the runtime silently convert — all layout transitions visible in the compiled IR.
  - **`stablehlo.custom_call "tt.sampling"` vs composite:** the op is registered as `torch.library.custom_op("tt::sampling")` in tt-xla and dispatched through XLA's `stablehlo_custom_call`, with no StableHLO decomposition. `custom_call` is therefore the appropriate lowering (unlike `tenstorrent.gather`/`tenstorrent.topk`, which have composite decompositions). Happy to
  revisit if there's a strong preference for a different form.

All Copilot feedback (description text, verifier rigor, batch==32 check) was already addressed in `41149c527`.

File: lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp (hunk 2 above)
Comment (kmabeeTT) [reply lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp]: Seemed doable, went ahead and added impl and test via https://github.com/tenstorrent/tt-mlir/pull/8043/commits/cf99c72016f582e18dcdc77a3213e1bce831671b pushed.

full `OpModel<SamplingOp>` implementation with 5-tensor `convertToTensorSpec` + `QUERY_OP_CONSTRAINTS(::ttnn::sampling, ...)`; `OpModelInterface` now wires `getOpConstraints`/`getOpRuntime` (replacing the `MissingMetalDefinition` stubs). Unit test `OpModelBase.SamplingOp` added in `TestOpModelInterface.cpp` (batch=32, candidates=128).

File: test/ttmlir/Dialect/TTNN/Transforms/Workarounds/sampling_workaround.mlir (hunk 3 above)
Comment (kmabeeTT) [reply test/ttmlir/Dialect/TTNN/Transforms/Workarounds/sampling_workaround.mlir:1]: Done at https://github.com/tenstorrent/tt-mlir/pull/8043/commits/85c269d685e83dc32b6645e3616ed03b9fc23fdc

`TTNNBuilder.sampling()` method + `test_sampling` (parametrized over Llama-3.x and OPT-125M vocab sizes). PCC check disabled since multinomial sampling is stochastic; the test verifies compile + device execute end-to-end.

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 4]
```diff
@@ -3836,6 +3836,30 @@ ::mlir::LogicalResult PagedUpdateCacheOp::verify() {
...
+  if (inputIndicesType.getRank() < 2) {
+    return emitOpError("input_indices must have at least 2 dimensions");
+  }
+
+  if (inputValuesType.getShape() != inputIndicesType.getShape()) {
+    return emitOpError(
+        "input_values and input_indices must have the same shape");
+  }
+
+  return success();
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply lib/Dialect/TTNN/IR/TTNNOps.cpp:3898]: resolved

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 5]
```diff
@@ -5224,6 +5224,30 @@ ::mlir::LogicalResult PagedUpdateCacheOp::verify() {
...
+  if (inputIndicesType.getRank() < 2) {
+    return emitOpError("input_indices must have at least 2 dimensions");
+  }
+
+  if (inputValuesType.getShape() != inputIndicesType.getShape()) {
+    return emitOpError(
+        "input_values and input_indices must have the same shape");
+  }
+
+  return success();
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:5277]: resolved

File: runtime/lib/ttnn/operations/reduction/sampling.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  // The compiler passes 2D [batch, candidates]. Reshape to 4D.
+  auto inputShape = inputValues.logical_shape();
+  if (inputShape.rank() == 2) {
+    uint32_t batch = inputShape[0];
+    uint32_t candidates = inputShape[1];
+    inputValues =
+        ::ttnn::reshape(inputValues, ::ttnn::Shape({1, 1, batch, candidates}));
+    inputIndices =
+        ::ttnn::reshape(inputIndices, ::ttnn::Shape({1, 1, batch, candidates}));
+  }
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply runtime/lib/ttnn/operations/reduction/sampling.cpp]: resolved

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 7]
```diff
@@ -4006,6 +4006,42 @@ def TTNN_AggregateTensorOp : TTNN_Op<"aggregate_tensor"> {
...
+    Inputs:
+      - `input_values`: Pre-filtered candidate logit values [batch, candidates] bf16, TILE layout
+      - `input_indices`: Global vocab indices for each candidate [batch, candidates] int32, ROW_MAJOR
+      - `k`: Per-request top-k values [32] uint32, ROW_MAJOR
+      - `p`: Per-request top-p values [32] bf16, ROW_MAJOR
+      - `temp`: Per-request temperature values [32] bf16, ROW_MAJOR
+      - `seed`: Optional random seed (uint32 scalar attribute)
+
+    Output:
+      - `result`: Sampled global token indices [1, 1, 1, batch] uint32, ROW_MAJOR
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td]: resolved

**Comment by kmabeeTT** [comment]
A couple test failures (d'oh I only had blackhole machine to test on locally originally), fixed now w/ follow up commit.

**Review by sdjordjevicTT — APPROVED** [review]
Thanks for addressing comments! All good from the compiler side.

**Review by sgholamiTT — APPROVED** [review]
OpModel changes LGTM. 👍🏻

**Review by pilkicTT — APPROVED** [review]

**Review by mrakitaTT — APPROVED** [review]

**Review by mmanzoorTT — APPROVED** [review]

File: tools/builder/ttnn/ttnn_builder.py [hunk 8]
```diff
@@ -9747,6 +9747,73 @@ def decorated_func(*inputs):
...
+        batch = vals_t.shape[0]
+        temperature = temp_t.clamp(min=1e-6).unsqueeze(-1)
+        scaled = torch.div(vals_t, temperature)
+        probs = torch.softmax(scaled, dim=-1)
+        sampled_local = torch.multinomial(probs, num_samples=1).squeeze(-1)
+        golden_output = (
+            torch.gather(idx_t, 1, sampled_local.unsqueeze(-1))
+            .squeeze(-1)
+            .to(torch.int32)
+        )
```
Comment (jgrimTT) [inline tools/builder/ttnn/ttnn_builder.py]: Can you move this golden handling to a function in the golden mappings (mapping.py)?

**Review by jgrimTT — CHANGES_REQUESTED** [review]
One change requested :)

File: tools/builder/ttnn/ttnn_builder.py (hunk 8 above)
Comment (kmabeeTT) [reply tools/builder/ttnn/ttnn_builder.py]: Done. Moved to `ttnn_sampling_golden` in `tools/golden/mapping.py` and registered in `GOLDEN_MAPPINGS`. Builder now fetches it via `get_golden_function(ttnn.SamplingOp)` like every other op. Unused `k`/`p`/`seed` params on the CPU side are underscore-prefixed following the `_num_experts_attr` convention in the same file.

Tests still passing:

```
test/python/golden/test_ttnn_ops.py::test_sampling[opt] PASSED                                                                                                                                                                                                                                                                                                   [ 50%]
test/python/golden/test_ttnn_ops.py::test_sampling[llama] PASSED
```

**Review by jgrimTT — APPROVED** [review]
